### ISSUE TITLE
Add an Effect(Model, Msg) option to Update enum

### ISSUE THREAD

**Issue by jesskfullwood** [David-OConnor/seed#52]
My initial view of this lib is "cool, Elm but in Rust!". However it has also elements of React (componentDidMount etc) which reduce the "purity" of the experience for some convenience, which is obviously a design tradeoff. 

Now in Elm, the `update` function returns a `Model` and a `Msg`. The model is immediately rendered by the `view` function, the `Msg` (if any) is added back into the event loop and is handled the next time round. This means data and effects flow in the same direction all the time, which IMO makes it nice to reason about.

The `update` function is `seed` is a little different. It only returns a `Model`, or rather, an `Update<Model>`. Any effects from the message (i.e. creating more messages) are, as far as I can tell, supposed to be handled 'in place', by passing around the `App` and running `App.update` as and when it is necessary.

I hit a 'race condition' of sorts using this style. I made a login `fetch` request, and when the function returns I want to redirect to the `Home` route. When `fetch` resolves it runs something like 
```
|login_data| app.update(Msg::LoggedIn(login_data, app.clone()))
```
This callback triggers a new message to be sent through the event loop, OK, fine. Then I handle it in the `update` function like
```
match msg {       
   LoggedIn(session, app) => {
        app.update(Msg::ChangePage(Route::Home));
        Model { session: Some(session), ..model }
    },
    // ....
}
```
Spot the bug? `app.update` runs immediately, redirecting to `Home`, then returns to the `update` function, which creates a new model with the previous `Route`, effectively taking me back to the login page. I could change the new model to point to Route::Home too, but I really want all routing to happen only through `Msg::ChangePage` as that handles url-updating as well.

Really what I want is the Elm behavior of just returning a variant `Effect(Model, Msg)` from my match statement, and have the framework handle it in order. The I won't have to pass the `App` around, also.

Thoughts?

**Comment by David-OConnor** [comment]
I never liked passing `App` around; would love to switch to the style you proposed.

**Comment by David-OConnor** [comment]
Check the latest commit. Basic functionality works as you describe, allowing chaining. Haven't found a way around passing the state/App around yet, but would ideally like to have a `seed::update` fn that can be called from anywhere.

**Comment by jesskfullwood** [comment]
Very cool! I look forward to trying this out

**Comment by David-OConnor** [comment]
Related, in earlier versions, the `update` function (user-created, not App.update) returned a model directly, allowing you to do recursive calls within it, ie:

```rust
match msg {       
   LoggedIn(session, app) => {
       model_ = update(Msg::ChangePage(Route::Home));
       Model { session: Some(session), ..model }
    },
    // ....
}
```

**Comment by jesskfullwood** [comment]
I tried it out and it tidied up lots of code which was very nice.

There is still a need to pass around the `App` in some places, because e.g. when the `login` button is clicked it sends a `LoginSubmit` message which needs the `App` to launch a `fetch` request.
```
LoginSubmit { data, app } => {
    fetch::spawn_local(login_request(data, app));
    Update::Skip(model)
}
```

One way around this would be to make `fetch::spawn_local` accept
```
Future<Item = Msg, Error = JsValue> + 'static
```
and the `Msg` could just be processed by `app.update` when it resolves (a la Elm) instead of needing to pass an `App` into the future. But I haven't looked at how this could be implemented.

Speaking in more general terms, it might make the execution model more consistent if each time `app.update` was called, it added the message to a queue, and then handled messages handled in FIFO order.

(BTW thanks for your hard work, think this framework has great potential)

**Comment by David-OConnor** [comment]
Agree on that fetch approach. I'm going to add an `Update` unwrapper to the next commit, leaving us with three related patterns of using multiple messages in the update fn:

The example I posted above, modified to work with this new unwrapper. Updates the model from one message, then modifies it further with the current one, rendering after both changes are made.
```rust
match msg {       
   LoggedIn(session, app) => {
       Render(
           Model { session: Some(session), ..update(Msg::ChangePage(Route::Home)).model()}
       )
    },
    // ....
}
```

The `Effect` variant: Similar to above, but renders after each update.
```rust
match msg {       
   LoggedIn(session, app) => {
        Effect(
            Model { session: Some(session), ..model }, 
            Msg::ChangePage(Route::Home)
        )
    },
    // ....
}
```
Executing side effects, then updating based on a different message:
```rust
Msg::RoutePage(page) => {
    seed::push_path(vec![&page.to_string()]);
    update(Msg::ChangePage(page), model)
},
```

**Comment by David-OConnor** [comment]
What do you think about using `Update::RenderThen` instead of `Update::Effect` ? Seems to be more descriptive of what it does.

**Comment by David-OConnor** [comment]
Published as v0.2.9

**Comment by jesskfullwood** [comment]
I finally got around to trying out the API, and it feels like a nice improvement. I have some thoughts about how to stop passing `App` around; I will start a new issue for that.
